### PR TITLE
Add OLE DB support

### PR DESCRIPTION
**Pull request by trevorbonas**
### Summary

Add OLE DB support

### Description

- Documentation for how the driver can be used with OLE DB, using PowerShell and Excel, has been added in its own file, `docs/markdown/setup/ole-db.md`.
- Null is now returned for null values instead of the character "`-`".
- Empty arrays are represented as `[]` instead of `[-]`.
- `byteOffset` changed to `size_t`.
- `columnBindOffset` changed to `SqlUlen*`.
- `SQLExtendedFetch` implemented separately from `SQLFetchScroll` to support ODBC 2.x applications.
- `rowsetSize` added to `Statement`, for use with `SQLExtendedFetch`.
- `SQL_ROWSET_SIZE` is now handled in `Statement::InternalSetAttribute` and `Statement::InternalGetAttribute`.
- An `ignoreErrors` argument has been added to `InternalClose` so that `SQLFreeStmt` can ignore all errors when freeing.
- Restriction for array size being less than 1000 has been removed.
- Spelling errors have been fixed in comments and in the documentation.
- `SQLGetData` can now retrieve variable-length data in parts.
- Wth `SQLGetData`, the indicator pointer now receives the entire size of the cell if all data was placed in the buffer and gets the remaining data if only part of the data was placed in the buffer, when the buffer is too small to fit all data.

### Related Issue

N/A

### Additional Reviewers

N/A

### CODE REVIEW

File: src/odbc/src/statement.cpp [hunk 1]
```diff
@@ -344,19 +344,25 @@ SqlResult::Type Statement::InternalSetAttribute(int attr, void* value,
...
     }
 
+    case SQL_ROWSET_SIZE: {
+      SqlUlen val = reinterpret_cast<SqlUlen>(value);
+
+      LOG_DEBUG_MSG("SQL_ROWSET_SIZE: " << val);
+
+      rowsetSize = val;
+
+      ard->GetHeader().rowsetSize = rowsetSize;
```
Comment (trevorbonas) [inline src/odbc/src/statement.cpp:354]: Would this require the ARD fields be restored later?

File: src/odbc/src/statement.cpp [hunk 2]
```diff
@@ -551,6 +551,18 @@ SqlResult::Type Statement::InternalGetAttribute(int attr, void* buf, SQLINTEGER,
       break;
     }
 
+    case SQL_ROWSET_SIZE: {
+      SQLINTEGER* val = reinterpret_cast<SQLINTEGER*>(buf);
```
Comment (jduo) [inline src/odbc/src/statement.cpp:555]: I think this is a SQLULEN but can't find a reference in the docs.

File: src/odbc/src/statement.cpp [hunk 3]
```diff
@@ -976,11 +989,11 @@ SqlResult::Type Statement::InternalFreeResources(int16_t option) {
...
 
 void Statement::Close() {
-  IGNITE_ODBC_API_CALL(InternalClose());
+  IGNITE_ODBC_API_CALL(InternalClose(false));
 }
 
-SqlResult::Type Statement::InternalClose() {
-  if (!currentQuery.get())
+SqlResult::Type Statement::InternalClose(bool ignoreErrors) {
+  if (ignoreErrors || !currentQuery.get())
```
Comment (jduo) [inline src/odbc/src/statement.cpp:996]: This shouldn't return so early. We're missing calling currentQuery->Close() if this happens.

File: src/odbc/src/statement.cpp [hunk 4]
```diff
@@ -994,6 +1007,73 @@ SqlResult::Type Statement::InternalClose() {
...
+    return SqlResult::AI_ERROR;
+  }
+
+  SqlUlen columnBindOffsetValue = columnBindOffset ? *columnBindOffset : 0;
+  for (app::ColumnBindingMap::iterator it = columnBindings.begin();
+    it != columnBindings.end(); ++it) {
+    it->second.SetByteOffset(columnBindOffsetValue);
+  }
+
+  SQLINTEGER fetched = 0;
```
Comment (jduo) [inline src/odbc/src/statement.cpp:1040]: This should be a SQLULEN

File: src/odbc/src/statement.cpp [hunk 5]
```diff
@@ -994,6 +1007,73 @@ SqlResult::Type Statement::InternalClose() {
...
+  if (rowCount)
+    *rowCount = 0;
+
+  if (!currentQuery.get()) {
+    std::string errMsg = "Cursor is not in the open state.";
+    AddStatusRecord(SqlState::S24000_INVALID_CURSOR_STATE, errMsg);
+    return SqlResult::AI_ERROR;
+  }
+
+  SqlUlen columnBindOffsetValue = columnBindOffset ? *columnBindOffset : 0;
```
Comment (jduo) [inline src/odbc/src/statement.cpp:1034]: This should not use columnBindOffset which comes from SQL_ATTR_BIND_OFFSET_PTR. It should explicitly set ByteOffset to zero since SQLExtendedFetch doesn't support bind offsets.

File: src/odbc/src/statement.cpp [hunk 6]
```diff
@@ -994,6 +1007,73 @@ SqlResult::Type Statement::InternalClose() {
...
+      ++errors;
+    }
+
+    if (rowStatusArray) {
+      rowStatusArray[i] = SqlResultToRowResult(res);
+    }
+  }
+
+  if (rowCount) {
+    *rowCount = fetched < 0 ? static_cast<SQLULEN>(rowsetSize) : fetched;
```
Comment (jduo) [inline src/odbc/src/statement.cpp:1064]: Why does this get set to rowsetSize if fetched < 0?

File: src/odbc/src/statement.cpp [hunk 7]
```diff
@@ -1029,19 +1109,18 @@ SqlResult::Type Statement::InternalFetchRow() {
     return SqlResult::AI_ERROR;
   }
 
-  if (columnBindOffset) {
-    for (app::ColumnBindingMap::iterator it = columnBindings.begin();
-         it != columnBindings.end(); ++it)
-      it->second.SetByteOffset(*columnBindOffset);
-  }
+  SqlUlen columnBindOffsetValue = columnBindOffset ? *columnBindOffset : 0;
```
Comment (jduo) [inline src/odbc/src/statement.cpp:1112]: Should probably leave a comment on why we don't just skip the loop if columnBindOffset is null.

File: src/odbc/src/statement.cpp [hunk 8]
```diff
@@ -47,9 +47,9 @@ Statement::Statement(Connection& parent)
     : connection(parent),
       columnBindings(),
       currentQuery(),
-      rowsFetched(0),
-      rowStatuses(0),
-      columnBindOffset(0),
+      rowsFetched(nullptr),
+      rowStatuses(nullptr),
+      columnBindOffset(nullptr),
```
Comment (jduo) [inline src/odbc/src/statement.cpp:52]: Need to add rowsetSize to the initializer list.

File: src/odbc/src/statement.cpp (hunk 2 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp]: It makes sense to me that this is a `SqlUint`. I have changed this and changed `valueLen` to be set to `SQL_IS_UINTEGER`.

File: src/odbc/src/statement.cpp (hunk 8 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp:52]: Added. I have followed the implementation for initializing `rowArraySize` to `1`, though I'm not entirely sure why that is done.

File: src/odbc/src/statement.cpp (hunk 3 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp:997]: Changed so that `currentQuery->Close()` will be called if `ignoreErrors` is passed in.

File: src/odbc/src/statement.cpp (hunk 4 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp]: Changed, along with `errors`.

File: src/odbc/src/statement.cpp (hunk 5 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp]: Changed so that each byte offset is set to `0` within the loop just below this line.

File: src/odbc/src/statement.cpp (hunk 6 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp]: This is copied from `InternalFetchRow`. It may be some kind of way to handle overflows, as `fetched` is a `SQLINTEGER` in `InternalFetchRow`.

I have changed this to simply set `*rowCount` to `fetched`.

File: src/odbc/src/statement.cpp (hunk 7 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp:1114]: Added.

File: src/odbc/src/statement.cpp (hunk 1 above)
Comment (jduo) [reply src/odbc/src/statement.cpp:355]: No, this is an explicit call to SQLSetStmtAttr

File: src/odbc/src/statement.cpp [hunk 9]
```diff
@@ -366,17 +373,11 @@ SqlResult::Type Statement::InternalSetAttribute(int attr, void* value,
       break;
     }
 
+    case SQL_ATTR_PARAM_BIND_OFFSET_PTR:
```
Comment (jduo) [inline src/odbc/src/statement.cpp:376]: This isn't related to parameters. Setting the param bind offset pointer should not affect the row bind offset ptr.

File: src/odbc/src/statement.cpp [hunk 10]
```diff
@@ -551,6 +552,18 @@ SqlResult::Type Statement::InternalGetAttribute(int attr, void* buf, SQLINTEGER,
       break;
     }
 
+    case SQL_ROWSET_SIZE: {
+      SqlUlen* val = reinterpret_cast<SqlUlen*>(buf);
```
Comment (jduo) [inline src/odbc/src/statement.cpp:556]: I think SQL_ATTR_ROW_ARRAY_SIZE is also supposed to be a SQLULEN instead of SQLINTEGER...

File: src/odbc/src/statement.cpp [hunk 11]
```diff
@@ -976,22 +990,88 @@ SqlResult::Type Statement::InternalFreeResources(int16_t option) {
...
     return SqlResult::AI_SUCCESS;
 
-  if (!currentQuery->DataAvailable()) {
+  if (!ignoreErrors && !currentQuery->DataAvailable()) {
     AddStatusRecord(SqlState::S24000_INVALID_CURSOR_STATE,
                     "No cursor was open");
     return SqlResult::AI_ERROR;
   }
 
   SqlResult::Type result = currentQuery->Close();
```
Comment (jduo) [inline src/odbc/src/statement.cpp:1006]: Is it safe to call this if DataAvailable() was false? In the previous version this would throw an error first, but now the error can get skipped.

File: src/odbc/src/statement.cpp [hunk 12]
```diff
@@ -976,22 +990,88 @@ SqlResult::Type Statement::InternalFreeResources(int16_t option) {
...
+  if (rowCount)
+    *rowCount = 0;
+
+  if (!currentQuery.get()) {
+    std::string errMsg = "Cursor is not in the open state.";
+    AddStatusRecord(SqlState::S24000_INVALID_CURSOR_STATE, errMsg);
+    return SqlResult::AI_ERROR;
+  }
+
+  for (app::ColumnBindingMap::iterator it = columnBindings.begin();
```
Comment (jduo) [inline src/odbc/src/statement.cpp:1035]: Let's add an explicit comment about why this is needed.

File: src/odbc/src/statement.cpp (hunk 9 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp]: I have changed this to simply setting the `ard->GetHeader().bindOffsetPtr`, as described in the [ODBC specification](https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlsetstmtattr-function?view=sql-server-ver16#statement-attributes):
> Setting this statement attribute sets the SQL_DESC_BIND_OFFSET_PTR field in the APD header.

It also says that `SQL_DESC_DATA_PTR`, `SQL_DESC_INDICATOR_PTR`, and `SQL_DESC_OCTET_LENGTH_PTR` should be set but we determined that we won't be doing that.

File: src/odbc/src/statement.cpp (hunk 10 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp:560]: Changed. I have also changed the data types of `fetched` and `errors` and removed the negative value check for `fetched` in `InternalFetchRow`.

File: src/odbc/src/statement.cpp (hunk 9 above)
Comment (jduo) [reply src/odbc/src/statement.cpp]: The spec is referring to the APD (Application Parameter Descriptor) which this driver doesn't really implement yet. The APD is for binding parameters for the application to push to the database through the driver.

You're setting it on the ARD (Application Row Descriptor) which is for binding buffers to write query results from SQL fetch.

File: src/odbc/src/statement.cpp (hunk 11 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp:1010]: Changed so `currentQuery->Close()` cannot be called after `currentQuery->DataAvailable()` is `false`. `ignoreErrors` now does what it sounds like, causes `SqlResult::AI_SUCCESS` to be returned no matter what and does not alter the logic of `InternalClose`.

File: src/odbc/src/statement.cpp [hunk 13]
```diff
@@ -434,7 +439,7 @@ void Statement::SetAttribute(StatementAttributes& stmtAttr) {
   SetAttribute(SQL_ATTR_CONCURRENCY, &stmtAttr.concurrency, 0);
   SetAttribute(SQL_ATTR_CURSOR_TYPE, &stmtAttr.cursorType, 0);
   SetAttribute(SQL_ATTR_RETRIEVE_DATA, &stmtAttr.retrievData, 0);
-  SetAttribute(SQL_ATTR_ROW_ARRAY_SIZE, &stmtAttr.rowsetSize, 0);
+  SetAttribute(SQL_ROWSET_SIZE, &stmtAttr.rowsetSize, 0);
```
Comment (jduo) [inline src/odbc/src/statement.cpp:442]: I'm not familiar enough with this codebase to tell if this is correct, but it seems odd. What is this function for?

File: src/odbc/src/statement.cpp (hunk 12 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp:1041]: Added.

File: src/odbc/src/statement.cpp (hunk 13 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp:442]: This version of `SetStatementAttribute` is only used in one place, in `connection.cpp`:
```
SqlResult::Type Connection::InternalCreateStatement(Statement*& statement) {
  LOG_DEBUG_MSG("InternalCreateStatement is called");
  statement = new Statement(*this);

  if (!statement) {
    AddStatusRecord(SqlState::SHY001_MEMORY_ALLOCATION, "Not enough memory.");

    return SqlResult::AI_ERROR;
  }

  // set statement attributes that are set to this connection
  statement->SetAttribute(stmtAttr_);
  return SqlResult::AI_SUCCESS;
}
```

`stmtAttr_` has its fields set in `Connection::InternalSetStmtAttribute`.

This seems to be a way for `Connection` to use `Statement`. I changed the implementation so that the `rowsetSize` value is handled as `SQL_ROWSET_SIZE` `Statement::InternalSetAttribute`, instead of being treated as `SQL_ATTR_ROW_ARRAY_SIZE`.

File: src/odbc/src/statement.cpp (hunk 9 above)
Comment (trevorbonas) [reply src/odbc/src/statement.cpp]: I see, my bad. I have changed this to set `apdi->GetHeader().bindOffsetPtr`, even though the driver doesn't currently use APD.

File: src/odbc/src/statement.cpp (hunk 13 above)
Comment (jduo) [reply src/odbc/src/statement.cpp:442]: OK this makes sense. Based on the properties being set they look like they are the properties in ODBC 2.x SQLSetStmtOption function so it likely relates to that.

File: src/odbc/include/timestream/odbc/statement.h [hunk 14]
```diff
@@ -65,19 +65,21 @@ class IGNITE_IMPORT_EXPORT Statement : public diagnostic::DiagnosableAdapter {
   void BindColumn(uint16_t columnIdx, int16_t targetType, void* targetValue,
                   SqlLen bufferLength, SqlLen* strLengthOrIndicator);
 
+  void Statement::ExtendedFetch(SQLUSMALLINT orientation, SQLLEN offset, SQLULEN* rowCount, SQLUSMALLINT* rowStatusArray);
```
Comment (jduo) [inline src/odbc/include/timestream/odbc/statement.h:68]: Please remove the Statement:: prefix here and in front of InternalExtendedFetch. It breaks non-Windows builds.

**Review by jduo — COMMENTED** [review]
This LGTM once you fixed the Statement:: prefix and add some end-to-end tests.

File: src/odbc/include/timestream/odbc/statement.h (hunk 14 above)
Comment (trevorbonas) [reply src/odbc/include/timestream/odbc/statement.h:68]: Removed.

**Comment by trevorbonas** [comment]
I have added a few integration tests. Also, an issue with my implementation of `SQLExtendedFetch` was revealed when running all other integration tests. I am working on fixing these issues.

File: src/odbc/src/statement.cpp (hunk 6 above)
Comment (jduo) [reply src/odbc/src/statement.cpp]: I would check over the logic in InternalFetchRow to make sure it makes sense.

File: src/odbc/src/odbc.cpp [hunk 15]
```diff
@@ -1182,7 +1185,18 @@ SQLRETURN SQLGetData(SQLHSTMT stmt, SQLUSMALLINT colNum, SQLSMALLINT targetType,
...
+  // between repeated calls on the same column.
+  if (colNum != statement->GetCurrentColNum()) {
+    statement->SetCurrentColNum(colNum);
+    statement->SetCellOffset(0);
+    dataBuffer.SetCellOffset(0);
+  } else {
+    dataBuffer.SetCellOffset(statement->GetCellOffset());
+  }
   statement->GetColumnData(colNum, dataBuffer);
+  statement->SetCellOffset(dataBuffer.GetCellOffset());
```
Comment (trevorbonas) [inline src/odbc/src/odbc.cpp:1199]: This is how `SQLGetData` can return variable-length data in parts. `Statement` persists across `SQLGetData` calls, `dataBuffer` does not. The cell offset is used within `dataBuffer` to calculate the number of bytes to skip in a field when writing to a buffer.

Please let me know if you have a better way of handling this. In its current form, The application data buffer's cell offset must be reset for every row fetch. This means tests using application data buffers have to call buffer.SetCellOffset(0) after every call involving putting a string to a buffer. This is not ideal.

Comment (trevorbonas) [reply src/odbc/src/odbc.cpp:1199]: Maybe a better way to do it would be with a pointer to a `Statement` cell offset. A new application data buffer constructor could be added like:
```
  ApplicationDataBuffer dataBuffer(driverType, targetValue, bufferLength,
                                   strLengthOrIndicator, &cellOffset);
```
Outside of its use in `SQLGetData`, the `cellOffsetPtr` in `ApplicationDataBuffer` could remain `nullptr`.

Comment (trevorbonas) [reply src/odbc/src/odbc.cpp:1199]: I have gone with a method where `cellOffset` is only used by `ApplicationDataBuffer` during `SQLGetData`. It is initialized to `-1` to indicate that it shouldn't be used.

`Statement` now only has to manage the cell offset in `SQLGetData` and in `InternalFetchRow`, resetting `cellOffset` to `0` for each new row.

File: src/tests/integration-test/src/queries_test.cpp [hunk 16]
```diff
@@ -1476,16 +1474,15 @@ BOOST_AUTO_TEST_CASE(TestDefaultValues) {
...
-  BOOST_CHECK_EQUAL(defaultBoolColumn, false);
-  BOOST_CHECK_EQUAL(defaultDoubleColumn, 0.0);
-  BOOST_CHECK_EQUAL(defaultBigintColumn, 0);
-  BOOST_CHECK_EQUAL(columnLens[3], 1);
-  BOOST_CHECK_EQUAL(columnLens[4], 8);
-  BOOST_CHECK_EQUAL(columnLens[5], 8);
+  BOOST_CHECK_EQUAL(doubleColumn, 63.7);
+  BOOST_CHECK_EQUAL(columnLens[0], 8);
+  BOOST_CHECK_EQUAL(columnLens[1], 16);
+  BOOST_CHECK_EQUAL(columnLens[2], 8);
```
Comment (trevorbonas) [inline src/tests/integration-test/src/queries_test.cpp:1485]: These changes are unrelated to the changes to add OLE DB support. The table in Timestream for LiveAnalytics simply has different data than what the test expects.

Comment (jduo) [reply src/tests/integration-test/src/queries_test.cpp:1485]: It might be worth generalizing tests to separate the data from the test execution. Then the tests can be applied to different data sets easily. (not in this PR).

**Review by jduo — CHANGES_REQUESTED** [review]

File: src/odbc/src/odbc.cpp [hunk 17]
```diff
@@ -1182,7 +1185,18 @@ SQLRETURN SQLGetData(SQLHSTMT stmt, SQLUSMALLINT colNum, SQLSMALLINT targetType,
   ApplicationDataBuffer dataBuffer(driverType, targetValue, bufferLength,
                                    strLengthOrIndicator);
 
+  // Cell offsets are used between SQLGetData calls to return
+  // variable-length data in parts. Statement keeps the cell offset
+  // between repeated calls on the same column.
+  if (colNum != statement->GetCurrentColNum()) {
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/odbc.cpp:1191]: `SQLFetch` will call `Statement::InternalFetchRow`. I have added a line to reset the `cellOffset` to `0` for `Statement` there.

`currentColNum` is only used in `SQLGetData`.

File: src/odbc/src/app/application_data_buffer.cpp [hunk 18]
```diff
@@ -42,7 +42,8 @@ ApplicationDataBuffer::ApplicationDataBuffer()
       buflen(0),
       reslen(0),
       byteOffset(0),
-      elementOffset(0) {
+      elementOffset(0),
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/app/application_data_buffer.cpp]: Fixed for `ApplicationDataBuffer` and `Statement`.

File: src/odbc/src/app/application_data_buffer.cpp [hunk 19]
```diff
@@ -251,25 +255,52 @@ ConversionResult::Type ApplicationDataBuffer::PutStrToStrBuffer(
   LOG_DEBUG_MSG("inCharSize is " << inCharSize << ", outCharSize is "
                                  << outCharSize << ", buflen is " << buflen);
 
+  size_t bytesRequired = 0;
+  if (ANSI_STRING_ONLY) {
+    bytesRequired = value.length() * outCharSize;
+  } else {
+    static std::wstring_convert< std::codecvt_utf8< wchar_t >, wchar_t >
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/app/application_data_buffer.cpp]: Changed this converter and the converter in `utility.cpp` to non-static. Visual Studio says the converter is 84 bytes.

File: src/odbc/src/app/application_data_buffer.cpp [hunk 20]
```diff
@@ -251,25 +255,52 @@ ConversionResult::Type ApplicationDataBuffer::PutStrToStrBuffer(
...
+    std::wstring inString = converter.from_bytes(value.c_str());
+    bytesRequired = inString.length() * outCharSize;
+  }
+
+  // Safety check since inCharSize and outCharSize will be denominators.
+  if (inCharSize == 0 || outCharSize == 0) {
+    LOG_ERROR_MSG(
+      "Unexpected conversion from unknown type string, inCharSize is "
+      << inCharSize << ", outCharSize is " << outCharSize);
+    assert(false);
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/app/application_data_buffer.cpp:273]: `outCharSize` will originate with the `type` in `ApplicationDataBuffer`. For `SQLGetData`, this will be the `targetType` value, after conversion with `ToDriverType`. `inCharSize` is internal.

The driver in general doesn't seem to throw many exceptions. In this case should I throw something?

File: src/odbc/src/utility.cpp [hunk 21]
```diff
@@ -38,55 +38,58 @@ size_t CopyUtf8StringToSqlCharString(const char* inBuffer, SQLCHAR* outBuffer,
...
+    return std::min(inBufferLenBytes, outBufferLenBytes);
   } else {
     // the inBuffer may contain unicode characters
     // Need to convert input string to wide-char to get the
     // length in characters - as well as get .narrow() to work, as expected
     // Otherwise, it would be impossible to safely determine the
     // output buffer length needed.
     static std::wstring_convert< std::codecvt_utf8< wchar_t >, wchar_t >
-        converter;
+      converter;
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/utility.cpp:74]: Changed to `thread_local` instead of `static`.

File: src/tests/integration-test/src/api_robustness_test.cpp [hunk 22]
```diff
@@ -812,6 +856,109 @@ BOOST_AUTO_TEST_CASE(TestSQLGetData) {
   SQLFetch(stmt);
 }
 
+BOOST_AUTO_TEST_CASE(TestSQLGetDataVarcharAsciiZeroBufferLength) {
+  // Ensures that calling SQLGetData with a buffer length of zero
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/tests/integration-test/src/api_robustness_test.cpp:860]: Do you mean when the buffer is NULL, i.e., the argument `TargetValuePointer`? [The SQLGetData documentation](https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlgetdata-function?view=sql-server-ver16#retrieving-data-with-sqlgetdata) says it cannot be NULL. When I tested, the ODBC manager returns "Invalid argument value."

File: src/odbc/src/app/application_data_buffer.cpp [hunk 23]
```diff
@@ -251,25 +255,52 @@ ConversionResult::Type ApplicationDataBuffer::PutStrToStrBuffer(
...
+  }
+
   SqlLen* resLenPtr = GetResLen();
   void* dataPtr = GetData();
 
-  if (!dataPtr)
+  if (!dataPtr) {
     return ConversionResult::Type::AI_SUCCESS;
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/app/application_data_buffer.cpp:284]: The `SQLGetData` specification says the input buffer cannot be a NULL pointer. I have made this change anyways.

File: src/odbc/src/connection.cpp [hunk 24]
```diff
@@ -84,6 +84,14 @@ Connection::Connection(Environment* env)
 
   // record the Connection object in an atomic counter
   ++refCount_;
+
+  // Set default values for stmtAttr_
+  stmtAttr_.bindType = SQL_PARAM_BIND_BY_COLUMN;
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/connection.cpp:89]: `StatementAttributes` is a struct. I have initialized `Connection::stmtAttr_` within `connection.h` using the following initializer list:
```c++
  /** statement attributes struct */
  StatementAttributes stmtAttr_ = {
    SQL_PARAM_BIND_BY_COLUMN, // bindType
    SQL_CONCUR_READ_ONLY,     // concurrency
    SQL_CURSOR_FORWARD_ONLY,  // cursorType
    SQL_RD_OFF,               // retrieveData
    1,                        // rowsetSize
    1,                        // rowArraySize
  };
```

File: src/odbc/src/statement.cpp [hunk 25]
```diff
@@ -551,13 +575,25 @@ SqlResult::Type Statement::InternalGetAttribute(int attr, void* buf, SQLINTEGER,
       break;
     }
 
+    case SQL_ROWSET_SIZE: {
+      SqlUlen* val = reinterpret_cast<SqlUlen*>(buf);
+
+      *val = rowsetSize;
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/statement.cpp]: It wasn't checked anywhere but doesn't seem to cause problems. There is even a test for it:
```c++
  SQLGetStmtAttr(stmt, SQL_ATTR_ROW_ARRAY_SIZE, 0, 0, 0);
```

I have added null checks just in case for all `val` values in `InternalGetAttribute`.

File: src/odbc/src/statement.cpp [hunk 26]
```diff
@@ -976,22 +1013,90 @@ SqlResult::Type Statement::InternalFreeResources(int16_t option) {
...
+  if (orientation != SQL_FETCH_NEXT) {
+    AddStatusRecord(SqlState::SHYC00_OPTIONAL_FEATURE_NOT_IMPLEMENTED,
+      "Only SQL_FETCH_NEXT FetchOrientation type is supported");
+    return SqlResult::AI_ERROR;
+  }
+
+  if (rowCount)
+    *rowCount = 0;
+
+  if (!currentQuery.get()) {
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/statement.cpp:1091]: I have placed the setting of `*rowCount` to `0` after checking whether the current query can be gotten.

File: src/odbc/src/statement.cpp [hunk 27]
```diff
@@ -976,22 +1013,90 @@ SqlResult::Type Statement::InternalFreeResources(int16_t option) {
...
+  if (rowCount) {
+    *rowCount = fetched;
+  }
+
+  if (fetched > 0) {
+    return errors == 0 ? SqlResult::AI_SUCCESS
+      : SqlResult::AI_SUCCESS_WITH_INFO;
+  }
+  
+  LOG_DEBUG_MSG("rowCount is " << rowCount << ", fetched is " << fetched
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/statement.cpp]: Changed to check `rowCount` and report `*rowCount`. I have also moved this debug log so that it is reported whether errors occur or not, by moving it above the `fetched` check.

File: src/odbc/src/utility.cpp [hunk 28]
```diff
@@ -38,55 +38,58 @@ size_t CopyUtf8StringToSqlCharString(const char* inBuffer, SQLCHAR* outBuffer,
...
+      }
     } else {
-      strncpy(reinterpret_cast< char* >(outBuffer), inBuffer, inBufLen);
-      outBuffer[inBufLen] = 0;
       isTruncated = false;
+      strncpy(reinterpret_cast< char* >(outBuffer), inBuffer, inBufferLenBytes);
+      outBuffer[inBufferLenBytes] = 0;
     }
-    return std::min(inBufLen, outBufferLenBytes - 1);
+    return std::min(inBufferLenBytes, outBufferLenBytes);
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/utility.cpp]: Yes, I have changed this back to how it was.

File: src/tests/integration-test/src/attributes_test.cpp [hunk 29]
```diff
@@ -290,22 +290,20 @@ BOOST_AUTO_TEST_CASE(StatementAttributeCursorType) {
...
-  // repeat test for different values
   SQLULEN val = 5;
   SQLRETURN ret =
       SQLSetStmtAttr(stmt, SQL_ATTR_ROW_ARRAY_SIZE,
                      reinterpret_cast< SQLPOINTER >(val), sizeof(val));
 
   ODBC_FAIL_ON_ERROR(ret, SQL_HANDLE_STMT, stmt);
 
-  ret = SQLGetStmtAttr(stmt, SQL_ATTR_ROW_ARRAY_SIZE, &actual_row_array_size,
+  ret = SQLGetStmtAttr(stmt, SQL_ATTR_ROW_ARRAY_SIZE, reinterpret_cast<SQLPOINTER>(&actual_row_array_size),
```
[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/tests/integration-test/src/attributes_test.cpp]: Removed.

File: src/odbc/src/app/application_data_buffer.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (RoyZhang2022) [reply src/odbc/src/app/application_data_buffer.cpp:273]: I think this if will never be reached as sizeof will never return 0 even for a empty class or struct.

[the remark this answers is not in the document]
Comment (trevorbonas) [reply src/odbc/src/app/application_data_buffer.cpp]: Good point. Removed.

File: src/odbc/src/statement.cpp [hunk 30]
```diff
@@ -455,7 +479,9 @@ SqlResult::Type Statement::InternalGetAttribute(int attr, void* buf, SQLINTEGER,
     case SQL_ATTR_APP_ROW_DESC: {
       SQLPOINTER* val = reinterpret_cast< SQLPOINTER* >(buf);
 
-      *val = static_cast< SQLPOINTER >(ard);
+      if (val) {
```
Comment (RoyZhang2022) [inline src/odbc/src/statement.cpp:482]: On above line 472 there is a check for buf nullptr. So no need to check here.

Comment (trevorbonas) [reply src/odbc/src/statement.cpp]: Removed.

**Review by RoyZhang2022 — APPROVED** [review]

File: docs/markdown/setup/ole-db.md [hunk 31]
```diff
@@ -0,0 +1,56 @@
+# OLE DB Setup and Support
+
+OLE DB can be used with the Amazon Timestream ODBC driver.
```
Comment (jduo) [inline docs/markdown/setup/ole-db.md:3]: [The OLE DB to ODBC bridge](https://learn.microsoft.com/en-us/sql/ado/guide/appendixes/microsoft-ole-db-provider-for-odbc?view=sql-server-ver15)](https://learn.microsoft.com/en-us/sql/ado/guide/appendixes/microsoft-ole-db-provider-for-odbc?view=sql-server-ver15) can be used with the Amazon Timestream ODBC driver.

**Review by jduo — APPROVED** [review]

File: docs/markdown/setup/ole-db.md (hunk 31 above)
Comment (trevorbonas) [reply docs/markdown/setup/ole-db.md]: Changed.
